File: src/lib/scala/Slate.Core/src/main/scala/slate/core/apis/ApiHelper.scala
/**
<slate_header>
  url: www.slatekit.com
  git: www.github.com/code-helix/slatekit
  org: www.codehelix.co
  author: Kishore Reddy
  copyright: 2016 CodeHelix Solutions Inc.
  license: refer to website and/or github
  about: A Scala utility library, tool-kit and server backend.
  mantra: Simplicity above all else
</slate_header>
  */

package slate.core.apis

import slate.common.results.ResultSupportIn

import scala.collection.mutable.Map
import slate.common.{InputArgs, Strings, Result}
import slate.core.apis.support.ApiCallReflect

object ApiHelper extends ResultSupportIn {

  def isValidMatch(actual:String, expected:String):Boolean = {
    if ( Strings.isNullOrEmpty(actual) || actual == "*" )
      true
    else
      Strings.isMatch(actual, expected)
  }


  def buildArgs(inputs:Option[List[(String,String)]]):InputArgs = {

    // fill args
    val rawArgs = Map[String,Any]()
    inputs.fold(Unit)( all => {
      all.foreach( input => rawArgs(input._1) = input._2)
      Unit
    })
    val args = new InputArgs(rawArgs)
    args
  }


  def buildCmd(path:String,
               inputs:Option[List[(String,String)]],
               headers:Option[List[(String,String)]]): Request = {

    val tokens = Strings.split(path, '.').toList
    val args = Some(buildArgs(inputs))
    val opts = Some(buildArgs(headers))
    val apiCmd = new Request(path, tokens, tokens(0), tokens(1), tokens(2), "get", args, opts)
    apiCmd
  }


  /**
   *  Checks the action and api to ensure the current request (cmd) is authorizated to
    *  make the call
   */
  def isAuthorizedForCall(cmd:Request, call:ApiCallReflect, auth:Option[ApiAuth]):Result[Boolean] =
  {
    val noAuth = !auth.isDefined

    // CASE 1: No auth for action
    if(noAuth && (call.action.roles == ApiConstants.RoleGuest || Strings.isNullOrEmpty(call.action.roles) )){
      ok()
    }
    // CASE 2: No auth for parent
    else if(noAuth && call.action.roles == ApiConstants.RoleParent
        && call.api.roles == ApiConstants.RoleGuest){
      ok()
    }
    // CASE 3: No auth and action requires roles!
    else if(noAuth){
      unAuthorized(msg = Some("Unable to authorize, authorization provider not set"))
    }
    else {
      // auth-mode, action roles, api roles
      auth.get.isAuthorized(cmd, call.api.auth, call.action.roles, call.api.roles)
    }
  }


  def getReferencedValue(primaryValue:String, parentValue:String) : String = {

    // Role!
    if(!Strings.isNullOrEmpty(primaryValue) ){
      if(Strings.isMatch(primaryValue, ApiConstants.RoleParent)){
        parentValue
      }
      else
        primaryValue
    }
    // Parent!
    else if(!Strings.isNullOrEmpty(parentValue)){
      parentValue
    }
    else
      ""
  }


  /**
   * copies the annotation taking into account the overrides
    *
    * @param ano
   * @param roles
   * @param auth
   * @param protocol
   * @return
   */
  def copyApiAnnotation(ano:Api,
                        roles:Option[String] = None,
                        auth:Option[String] = None,
                        protocol:Option[String] = None ): Api = {
    if(!roles.isDefined && !auth.isDefined && !protocol.isDefined ){
      ano
    }
    else {
      val finalRoles = roles.getOrElse(ano.roles)
      val finalAuth = auth.getOrElse(ano.auth)
      val finalProtocol = protocol.getOrElse(ano.protocol)
      ano.copy(ano.area, ano.name, ano.desc, finalRoles, finalAuth, ano.verb, finalProtocol)
    }
  }
}

File: src/lib/kotlin/slatekit-apis/src/main/kotlin/slatekit/apis/helpers/ApiHelper.kt
/**
 * <slate_header>
 * url: www.slatekit.com
 * git: www.github.com/code-helix/slatekit
 * org: www.codehelix.co
 * author: Kishore Reddy
 * copyright: 2016 CodeHelix Solutions Inc.
 * license: refer to website and/or github
 * about: A tool-kit, utility library and server-backend
 * mantra: Simplicity above all else
 * </slate_header>
 */

package slatekit.apis.helpers

import slatekit.apis.*
import slatekit.apis.core.Target
import slatekit.apis.core.Api
import slatekit.apis.core.Auth
import slatekit.common.*
import slatekit.common.auth.AuthFuncs
import slatekit.common.auth.Roles
import slatekit.common.requests.InputArgs
import slatekit.common.requests.Request
import slatekit.common.CommonRequest
import slatekit.common.requests.Source
import slatekit.meta.Deserializer
import slatekit.results.Notice
import slatekit.results.Success
import slatekit.results.builders.Notices

object ApiHelper {

    val typeDefaults = mapOf(
            "String" to "",
            "Boolean" to false,
            "Int" to 0,
            "Long" to 0L,
            "Double" to 0.0,
            "DateTime" to DateTime.now()
    )

    /**
     * String match factoring in the wildcard "*"
     */
    fun isValidMatch(actual: String, expected: String): Boolean {
        return if (actual.isNullOrEmpty() || actual == "*")
            true
        else
            actual == expected
    }

    /**
     * Builds arguments supplied into the Inputs class which
     * is the base class for inputs in the request.
     */
    fun buildArgs(inputs: List<Pair<String, Any>>?): InputArgs {

        // fill args
        val rawArgs = inputs?.let { all -> all.toMap() } ?: mapOf()
        val args = InputArgs(rawArgs)
        return args
    }

    /**
     * builds the request
     */
    fun buildCliRequest(
        path: String,
        inputs: List<Pair<String, Any>>?,
        headers: List<Pair<String, Any>>?
    ): Request {

        val tokens = path.split('.').toList()
        val args = buildArgs(inputs)
        val opts = buildArgs(headers)
        val apiCmd = CommonRequest(path, tokens, Source.CLI, "get", args, opts,
                null, "", "", ApiConstants.version, DateTime.now())
        return apiCmd
    }


    fun fillArgs(deserializer: Deserializer, apiRef: Target, cmd: Request): Array<Any?> {
        val action = apiRef.action
        // Check 1: No args ?
        return if (!action.hasArgs)
            arrayOf()
        // Check 2: 1 param with default and no args
        else if (action.isSingleDefaultedArg() && cmd.data.size() == 0) {
            val argType = action.paramsUser[0].type.toString()
            val defaultVal = if (typeDefaults.contains(argType)) typeDefaults[argType] else null
            arrayOf<Any?>(defaultVal ?: "")
        } else {
            deserializer.deserialize(action.params)
        }
    }

    /**
     * copies the annotation taking into account the overrides
     *
     * @param ano
     * @param roles
     * @param auth
     * @param protocol
     * @return
     */
    fun buildApiInfo(ano: Api, reg: Api): Api {

        val finalRoles = reg.roles.nonEmptyOrDefault(ano.roles)
        val finalAuth = reg.auth.nonEmptyOrDefault(ano.auth)
        val finalProtocol = reg.protocol.nonEmptyOrDefault(ano.protocol)
        return reg.copy(
                area = ano.area,
                name = ano.name,
                desc = ano.desc,
                roles = finalRoles,
                auth = finalAuth,
                verb = ano.verb,
                protocol = finalProtocol
        )
    }

    /**
     * copies the annotation taking into account the overrides
     *
     * @param ano
     * @param roles
     * @param auth
     * @param protocol
     * @return
     */
    fun buildApiInfo(reg: Api): Api {

        val name = reg.cls.simpleName!!.removeSuffix("Controller").removeSuffix("Api").removeSuffix("API")
        val finalArea = reg.area.nonEmptyOrDefault("")
        val finalName = reg.name.nonEmptyOrDefault(name)
        val finalRoles = reg.roles.nonEmptyOrDefault("?")
        val finalAuth = reg.auth.nonEmptyOrDefault(ApiConstants.AuthModeAppKey)
        val finalProtocol = reg.protocol.nonEmptyOrDefault("*")
        val finalVerb = reg.verb.nonEmptyOrDefault("*")
        return reg.copy(
                area = finalArea,
                name = finalName,
                roles = finalRoles,
                auth = finalAuth,
                verb = finalVerb,
                protocol = finalProtocol
        )
    }
}
